common/src/main/scala/com/gu/recipeasy/models/RecipeReadiness.scala

```diff
@@ -11,17 +19,142 @@ object RecipeReadiness {
   def migrateNigelSlaterRecipes(db: DB): Int = {
     val recipes: List[Recipe] = selectNigelSlaterRecipes(db).filter(_.status != Impossible)
     recipes.foreach(recipe => db.setOriginalRecipeStatus(recipe.id, Impossible))
     recipes.size
   }
+
+  // -------------------------------------------------
+  // Stage 3 Support Functions
+
+  def recipeHasACuratedRecordWithNonEmptyIngredientsLists(db: DB, recipe: Recipe): Boolean = {
+    val maybeCuratedRecipe: Option[CuratedRecipeDB] = db.getCuratedRecipeByRecipeId(recipe.id)
+    maybeCuratedRecipe.exists(curatedRecipe => curatedRecipe.ingredientsLists.lists.size > 0)
+  }
+
+  private object ShortListItem {
+    private def isShort(length: Int) = length > 0 && length < 120
+    def matches(node: Node): Boolean = node match {
+      case tn: TextNode => isShort(tn.text.trim.size)
+      case elem: Element if (elem.tag.getName == "strong" || elem.tag.getName == "em") => { isShort(elem.text.trim.size) }
+      case other => false
+    }
+  }
+
+  private object ParaWithListOfShortTexts {
+    def matches(el: Element): Boolean = {
+      if (el == null) false
+      else {
+        if (el.tag.getName == "p") {
+          val pairs = el.childNodes.asScala.toList.grouped(2)
+          pairs.forall {
+            case x :: y :: Nil => ShortListItem.matches(x) && y.nodeName == "br"
+            case x :: Nil => ShortListItem.matches(x) // final element
+            case _ => false
+          }
+        } else false
+      }
+    }
+    def unapply(el: Element): Option[Element] = if (matches(el)) Some(el) else None
+  }
+  def buildIngredientList(para: Element): Option[IngredientsList] = {
+    def text(node: Node): String = node match {
+      case tn: TextNode => tn.text.trim
+      case elem: Element => elem.text.trim
+      case other => ""
+    }
+    val listItems: Seq[Node] = para.childNodes.asScala.filterNot(_.nodeName == "br")
+    val withoutServingCount = listItems.filterNot(n => text(n).toLowerCase.dropWhile(_ == '(').startsWith("serves"))
+    if (withoutServingCount.size < 2) None
+    else {
+      if (withoutServingCount(0).nodeName != withoutServingCount(1).nodeName) {
+        val title = text(withoutServingCount(0))
+        val ingredients = withoutServingCount.drop(1).map(text)
+        Some(IngredientsList(title = Some(title), ingredients))
+      } else {
+        val ingredients = withoutServingCount.map(text)
+        Some(IngredientsList(title = None, ingredients))
+      }
+    }
+  }
+  def guessIngredients(body: Seq[Element]): Seq[IngredientsList] = {
+    // Find paragraphs containing short text items separated by <br>
+    val candidates = body.filter(ParaWithListOfShortTexts.matches)
+    candidates.flatMap(buildIngredientList)
+  }
+
+  def extractIngredientsLists(recipe: Recipe): Seq[IngredientsList] = {
+    val doc = Jsoup.parse(recipe.body)
+    val body: Seq[Element] = doc.body.children.asScala
+    val ingredientsLists: Seq[IngredientsList] = guessIngredients(body)
```

Comment:
True. (In the process of writing the code I wanted to indicate to myself what `guessIngredients` was returning, but this is actually the return value of the function `extractIngredientsLists` itself).

Done. See coming update.